runelite-client/src/main/java/net/runelite/client/plugins/adventurelog/AdventureLogPlugin.java

```diff
@@ -0,0 +1,729 @@
+				if (currentBossData.getMonster().equals(eventBossData.getMonster())
+					&& compareWorldTypes(bossWorldTypes, currentWorldtypes))
+				{
+					currentBossData.incrementAmount();
+					timeSinceLastBossKill = Instant.now();
+				}
+				else
+				{
+					queueCurrentBoss();
+
+					currentBossData = eventBossData;
+					bossWorldTypes = currentWorldtypes;
+					timeSinceLastBossKill = Instant.now();
+				}
+			}
+		}
+		else
+		{
+			if (currentBossData != null)
+			{
+				queueCurrentBoss();
+			}
+
+			AdventureLogRecord record = new AdventureLogRecord(
+				event,
+				trimWorldType(WorldUtil.runeliteToHttpWorldTypes(client.getWorldType()))
+			);
+			queueLog(record);
+		}
+	}
+
+	@Subscribe
+	public void onGameStateChanged(GameStateChanged event)
+	{
+		if (event.getGameState() == GameState.LOADING)
+		{
+			return;
+		}
+
+		// first LOGGED_IN gamestate the player logs in
+		if (event.getGameState() == GameState.LOGGED_IN
+			&& (previousGameState == GameState.LOGIN_SCREEN
+			|| previousGameState == GameState.LOGIN_SCREEN_AUTHENTICATOR))
+		{
+			isFirstTick = true;
+		}
+		else if (event.getGameState() == GameState.LOGIN_SCREEN)
+		{
+			queueCurrentBoss();
+			submitLogs();
+			hashedUsername = null;
+		}
+
+		previousGameState = event.getGameState();
+	}
+
+	@Subscribe
+	public void onWidgetLoaded(WidgetLoaded event)
+	{
+		switch (event.getGroupId())
+		{
+			case (WidgetID.QUEST_COMPLETED_GROUP_ID):
+				// widget gets loaded prior to the text being set, so wait until the next tick
+				shouldCheckQuestWidget = true;
+				break;
+			case (WidgetID.CLUE_SCROLL_REWARD_GROUP_ID):
+				// clue difficulty should be set via ChatMessage for clue scrolls.
+				// Clue Scrolls use same InventoryID as Barrows
+				if (clueDifficulty == null)
+				{
+					return;
+				}
+
+				ClueData clueData = new ClueData(
+					clueDifficulty,
+					client.getItemContainer(InventoryID.BARROWS_REWARD).getItems()
+				);
+				eventBus.post(clueData);
+				break;
+			case (WidgetID.DIALOG_SPRITE_GROUP_ID):
+				shouldCheckCastleWarsWidget = true;
+				break;
+			case (WidgetID.DIALOG_NPC_GROUP_ID):
+				// This widget is loaded twice for whatever reason, this is the cooldown
+				if (shouldCheckPestControlWidget == 0)
+				{
+					shouldCheckPestControlWidget = 3;
+					System.out.println("WIDGET LOADED");
+				}
+				break;
+			case (WidgetID.DIALOG_TEXT_GROUP_ID):
+				shouldCheckDiary = true;
+				break;
+			case (WidgetID.BA_REWARD_GROUP_ID):
+				shouldCheckBarbAssaultWidget = true;
+				break;
+		}
+	}
+
+	@Subscribe
+	public void onGameTick(GameTick event)
+	{
+		checkQuestWidget();
+		checkCastleWarsWidget();
+		checkPestControlWidget();
+		checkDiary();
+		checkBarbAssault();
+	}
+
+	@Subscribe
+	public void onChatMessage(ChatMessage event)
+	{
+		String message = Text.removeTags(event.getMessage());
+		if (message.contains(PET_FOLLOWING_MESSAGE))
+		{
+			shouldCheckPetFollowing = true;
```

Comment:
You can probably just check the 16 msb in varp 447 for the follower npcId here instead which would let you skip the npcSpawned check later. Might be more relieable if the npcSpawned event would ever fire before this, but idk if it ever does. 
```
int petID = (client.getVar(VarPlayer.FOLLOWER) >> 16) & 0xffff;
```